engine/access/rpc/connection/manager.go

```diff
@@ -0,0 +1,226 @@
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/module"
+)
+
+// DefaultClientTimeout is used when making a GRPC request to a collection node or an execution node.
+const DefaultClientTimeout = 3 * time.Second
+
+// Manager provides methods for getting and managing gRPC client connections.
+type Manager struct {
+	cache      *Cache
+	logger     zerolog.Logger
+	metrics    module.AccessMetrics
+	maxMsgSize uint
+}
+
+// NewManager creates a new Manager with the specified parameters.
+func NewManager(
+	cache *Cache,
+	logger zerolog.Logger,
+	metrics module.AccessMetrics,
+	maxMsgSize uint,
+) Manager {
+	return Manager{
+		cache:      cache,
+		logger:     logger,
+		metrics:    metrics,
+		maxMsgSize: maxMsgSize,
+	}
+}
+
+// GetConnection returns a gRPC client connection for the given grpcAddress and timeout.
+// If a cache is used, it retrieves a cached connection, otherwise creates a new connection.
+// It returns the client connection and an io.Closer to close the connection when done.
+func (m *Manager) GetConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, io.Closer, error) {
+	if m.cache != nil {
+		conn, err := m.retrieveConnection(grpcAddress, timeout)
+		if err != nil {
+			return nil, nil, err
+		}
+		return conn, &noopCloser{}, err
+	}
+
+	conn, err := m.createConnection(grpcAddress, timeout, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return conn, io.Closer(conn), nil
+}
+
+// Remove removes the gRPC client connection associated with the given grpcAddress from the cache.
+// It returns true if the connection was removed successfully, false otherwise.
+func (m *Manager) Remove(grpcAddress string) bool {
+	if m.cache == nil {
+		return false
+	}
+
+	res, ok := m.cache.Get(grpcAddress)
+	if !ok {
+		return false
+	}
+
+	if !m.cache.Remove(grpcAddress) {
+		return false
+	}
+
+	// Close the connection only if it is successfully removed from the cache
+	res.Close()
+	return true
+}
+
+// HasCache returns true if the Manager has a cache, false otherwise.
+func (m *Manager) HasCache() bool {
+	return m.cache != nil
+}
+
+// retrieveConnection retrieves the CachedClient for the given grpcAddress from the cache or adds a new one if not present.
+// If the connection is already cached, it waits for the lock and returns the connection from the cache.
+// Otherwise, it creates a new connection and caches it.
+func (m *Manager) retrieveConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, error) {
+	client, ok := m.cache.GetOrAdd(grpcAddress, timeout)
+	if ok {
+		// The client was retrieved from the cache, wait for the lock
+		client.mu.Lock()
+		if m.metrics != nil {
+			m.metrics.ConnectionFromPoolReused()
+		}
+	} else {
+		// The client is new, add it to the cache
+		if m.metrics != nil {
+			m.metrics.ConnectionAddedToPool()
+		}
+	}
+	defer client.mu.Unlock()
+
+	if client.ClientConn != nil && client.ClientConn.GetState() != connectivity.Shutdown {
```

Comment:
Do we actually need check for state? We always remove connection from cache when we close the connection. 
Additionally, I would be more explicit with implementation of this function by implementing next logic:

1. `GetOrAdd` returns new client - we perform initialization of connection, record metrics and return.
2. `GetOrAdd` returns cached client - we take lock, record metrics and return

We can simplify and structure this function in a way that we have specific paths.
In reality it's the way I have described it, but the check for path is not based on having the client cached but rather on availability of `ClientConn`.

Reply:
@peterargue

Reply:
On other hand current logic helps to recover from case where someone closed the connection using `grpc.ClientConn.Close` instead of `Manager.Remove(grpcAddress)`, that is a good thing, but also encourages wrong usage of connection manager.

Reply:
Yea, this is also covers the case where the remote end hung up. I think it's a good idea to keep this, though if we didn't check, the application logic would likely handle it gracefully.